File: lib/model/subject.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'subject.freezed.dart';
part 'subject.g.dart';

@freezed
class Subject with _$Subject {
  const factory Subject({
    required String id,
    required String url,
    required String name,
    String? area,
    String? required,
    String? term,
    int? units,
    String? grade,
    String? staff,
    String? room,
  }) = _Subject;

  factory Subject.fromJson(Map<String, dynamic> json) =>
      _$SubjectFromJson(json);
}

File: schemas/subject.ts
import { z } from 'zod';

export const Subject = z.object({
  id: z.string(),
  url: z.string(),
  name: z.string(),
  area: z.string().nullable().optional(),
  required: z.string().nullable().optional(),
  term: z.string().nullable().optional(),
  units: z.string().nullable().optional(),
  grade: z.string().nullable().optional(),
  staff: z.string().nullable().optional(),
  room: z.string().nullable().optional(),
});

export type Subject = z.infer<typeof Subject>;
